### PR TITLE
Add a deprecation notice to shadow replicas 

### PR DESCRIPTION
**Pull request by bleskes**
Relates to #22024

On top of documentation, the PR adds deprecation loggers and deals with the resulting warning headers.

The yaml test is set exclude versions up to 6.0. This is need to make sure bwc tests pass until this is backported to 5.2.0 . Once that's done, I will change the yaml test version limits

### CODE REVIEW

**Review by dakrone — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 1]
```diff
@@ -1217,6 +1218,7 @@ public static IndexMetaData fromXContent(XContentParser parser) throws IOExcepti
      * {@link #isIndexUsingShadowReplicas(org.elasticsearch.common.settings.Settings)}.
      */
     public static boolean isOnSharedFilesystem(Settings settings) {
+        // don't use the settings directly, not to trigger manny deprecation logging
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Nits: `settings` -> `setting` and `not to trigger manny` -> `to not trigger verbose` (it doesn't have to be exactly this, but at least fix `manny`)

File: core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java [hunk 2]
```diff
@@ -1226,6 +1228,7 @@ public static boolean isOnSharedFilesystem(Settings settings) {
      * setting for this is <code>false</code>.
      */
     public static boolean isIndexUsingShadowReplicas(Settings settings) {
+        // don't use the settings directly, not to trigger manny deprecation logging
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java]: Nits: `settings` -> `setting` and `not to trigger manny` -> `to not trigger verbose` (it doesn't have to be exactly this, but at least fix `manny`)

File: core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java [hunk 3]
```diff
@@ -476,8 +476,10 @@ static NodeShardsResult buildVersionBasedNodeShardsResult(ShardRouting shard, bo
      * recovered on any node
      */
     private boolean recoverOnAnyNode(IndexMetaData metaData) {
+        // don't use the settings directly, not to trigger manny deprecation logging
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java]: Nits: `settings` -> `setting` and `not to trigger manny` -> `to not trigger verbose` (it doesn't have to be exactly this, but at least fix `manny`)

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java [hunk 4]
```diff
@@ -74,6 +74,16 @@
...
     private TestAllocator testAllocator;
 
+
+    /**
+     * needed due to random usage of {@link IndexMetaData#INDEX_SHARED_FS_ALLOW_RECOVERY_ON_ANY_NODE_SETTING}. removed once
+     * shadow replicas are removed.
+     */
+    @Override
+    protected boolean enableWarningsCheck() {
+        return false;
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Would you please add an assert guarding against this staying around after shadow replicas are removed?

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 5]
```diff
@@ -76,6 +76,15 @@
 
     private final ClusterStateChanges cluster = new ClusterStateChanges(xContentRegistry());
 
+    /**
+     * needed due to random usage of {@link IndexMetaData#INDEX_SHADOW_REPLICAS_SETTING}. removed once
+     * shadow replicas are removed.
+     */
+    @Override
+    protected boolean enableWarningsCheck() {
+        return false;
```
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: Would you please add an assert guarding against this staying around after shadow replicas are removed?

File: docs/reference/indices/shadow-replicas.asciidoc [hunk 6]
```diff
@@ -1,7 +1,7 @@
 [[indices-shadow-replicas]]
 == Shadow replica indices
 
-experimental[]
+deprecated[5.2.0, Shadow replicas don't see much usage and we are planning to remove them]
```
Comment (jasontedor) [inline docs/reference/indices/shadow-replicas.asciidoc]: "Planning" or "going"?

File: docs/reference/migration/migrate_6_0/indices.asciidoc [hunk 7]
```diff
@@ -27,3 +27,8 @@ PUT _template/template_2
 }
 --------------------------------------------------
 // CONSOLE
+
+
+=== Shadow Replicas are deprecated
+
+<<indices-shadow-replicas,Shadow Replicas>> don't see much usage and we are planning to remove them.
```
Comment (jasontedor) [inline docs/reference/migration/migrate_6_0/indices.asciidoc]: "Planning" or "going"?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left some nits.

**Comment by jasontedor** [comment]
Thanks @bleskes. Will you open a follow-up issue for `path.shared_data`?

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 4 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: I did :) that's why the java docs links to the setting in question

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java (hunk 5 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: same as above

File: docs/reference/indices/shadow-replicas.asciidoc (hunk 6 above)
Comment (bleskes) [inline docs/reference/indices/shadow-replicas.asciidoc]: I don't know. Chose for a softer language here.

File: core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java (hunk 4 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java]: Okay, I'm fine with that.

File: core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java (hunk 5 above)
Comment (jasontedor) [inline core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: Yes, I'm also fine with it here.

**Comment by bleskes** [comment]
@jasontedor I pushed another commit. Can you look?

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by bleskes** [comment]
retest this please

**Comment by bleskes** [comment]
Retest this please

**Comment by bleskes** [comment]
retest this please

**Pull request merged**